### PR TITLE
[KBSWITCH] Delete SHLoadRegUIString hack

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Remove the needless hack and reduce the maintenance cost. Because `shlwapi!SHLoadRegUIString` function is already implemented, so we don't need the hack any more.

JIRA issue: [CORE-10667](https://jira.reactos.org/browse/CORE-10667)

## Proposed changes

- Directly use `shlwapi!SHLoadRegUIString` function instead of a hack.
- Add link to `shlwapi.dll`.

### CODE REVIEW

File: base/applications/kbswitch/kbswitch.c [hunk 1]
```diff
@@ -92,15 +92,17 @@ GetSystemLibraryPath(LPTSTR szPath, SIZE_T cchPath, LPCTSTR FileName)
...
+/* FIXME: Use <shlwapi_undoc.h> and delete this */
+HRESULT WINAPI SHLoadRegUIStringW(HKEY hkey, LPCWSTR value, LPWSTR buf, DWORD size);
+
 static BOOL
 GetLayoutName(LPCTSTR szLayoutNum, LPTSTR szName, SIZE_T NameLength)
 {
     HKEY hKey;
     DWORD dwBufLen;
-    TCHAR szBuf[MAX_PATH], szDispName[MAX_PATH], szIndex[MAX_PATH], szPath[MAX_PATH];
+    TCHAR szBuf[MAX_PATH], szDispName[MAX_PATH];
```
Comment (HBelusca) [inline base/applications/kbswitch/kbswitch.c:103]:
```suggestion
    TCHAR szBuf[MAX_PATH];
    WCHAR szDispName[MAX_PATH];
```
szDispName is used with explicilty unicode SHLoadRegUIStringW.

File: base/applications/kbswitch/kbswitch.c [hunk 2]
```diff
@@ -114,45 +116,11 @@ GetLayoutName(LPCTSTR szLayoutNum, LPTSTR szName, SIZE_T NameLength)
...
-            }
-
-            if (ExpandEnvironmentStrings(szDispName, szPath, ARRAYSIZE(szPath)))
-            {
-                hLib = LoadLibrary(szPath);
-                if (hLib)
-                {
-                    if (LoadString(hLib, _ttoi(szIndex), szPath, ARRAYSIZE(szPath)))
-                    {
-                        StringCchCopy(szName, NameLength, szPath);
```
Comment (HBelusca) [inline base/applications/kbswitch/kbswitch.c:147]: This copy into the user-given szName hasn't been done in the new code. This needs to be fixed.

File: base/applications/kbswitch/kbswitch.c [hunk 3]
```diff
@@ -92,15 +92,17 @@ GetSystemLibraryPath(LPTSTR szPath, SIZE_T cchPath, LPCTSTR FileName)
...
 GetLayoutName(LPCTSTR szLayoutNum, LPTSTR szName, SIZE_T NameLength)
 {
     HKEY hKey;
     DWORD dwBufLen;
-    TCHAR szBuf[MAX_PATH], szDispName[MAX_PATH], szIndex[MAX_PATH], szPath[MAX_PATH];
+    TCHAR szBuf[MAX_PATH], szDispName[MAX_PATH];
     TCHAR szLCID[CCH_LAYOUT_ID + 1];
-    HANDLE hLib;
-    UINT i, j, k;
+    HRESULT hr;
```
Comment (HBelusca) [inline base/applications/kbswitch/kbswitch.c:105]:
```suggestion
```

File: base/applications/kbswitch/kbswitch.c [hunk 4]
```diff
@@ -92,15 +92,17 @@ GetSystemLibraryPath(LPTSTR szPath, SIZE_T cchPath, LPCTSTR FileName)
...
 }
 
+/* FIXME: Use <shlwapi_undoc.h> and delete this */
+HRESULT WINAPI SHLoadRegUIStringW(HKEY hkey, LPCWSTR value, LPWSTR buf, DWORD size);
+
 static BOOL
 GetLayoutName(LPCTSTR szLayoutNum, LPTSTR szName, SIZE_T NameLength)
 {
     HKEY hKey;
     DWORD dwBufLen;
```
Comment (HBelusca) [inline base/applications/kbswitch/kbswitch.c:102]:
```suggestion
    HRESULT hr;
    DWORD dwBufLen;
```

File: base/applications/kbswitch/kbswitch.c [hunk 5]
```diff
@@ -92,15 +92,17 @@ GetSystemLibraryPath(LPTSTR szPath, SIZE_T cchPath, LPCTSTR FileName)
     return TRUE;
 }
 
+/* FIXME: Use <shlwapi_undoc.h> and delete this */
+HRESULT WINAPI SHLoadRegUIStringW(HKEY hkey, LPCWSTR value, LPWSTR buf, DWORD size);
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply base/applications/kbswitch/kbswitch.c]: Fixed header and source.

File: base/applications/kbswitch/kbswitch.c (hunk 2 above)
Comment (katahiromz) [reply base/applications/kbswitch/kbswitch.c:147]: Fixed.

File: sdk/include/reactos/shlwapi_undoc.h [hunk 6]
```diff
@@ -21,6 +21,11 @@
 #ifndef __SHLWAPI_UNDOC_H
 #define __SHLWAPI_UNDOC_H
 
+#include <unknwn.h>
+#include <docobj.h>
+#include <shlobj.h>
+#include <shobjidl.h>
```
Comment (HBelusca) [inline sdk/include/reactos/shlwapi_undoc.h:27]: Why all those headers? They should be included instead in the module that also includes shlwapi_undoc.h

Comment (katahiromz) [reply sdk/include/reactos/shlwapi_undoc.h]: Reverted.

File: base/applications/kbswitch/kbswitch.c [hunk 7]
```diff
@@ -8,6 +8,11 @@
  */
 
 #include "kbswitch.h"
+#include <unknwn.h>
+#include <docobj.h>
+#include <shlobj.h>
+#include <shobjidl.h>
```
Comment (HBelusca) [inline base/applications/kbswitch/kbswitch.c:14]: Are *all* those headers necessary? Can we remove some of these? (For example, are the first two "unknwn.h" and "docobj.h" necessary?)

**Review by HBelusca — APPROVED** [review]

File: base/applications/kbswitch/kbswitch.c (hunk 7 above)
Comment (katahiromz) [reply base/applications/kbswitch/kbswitch.c]: Deleted some `#include`'s.

**Pull request merged**